### PR TITLE
Basic WebUSB support (mobile)

### PR DESCRIPTION
**Pull request by slimonslimon**
WebUSB api will be used as fallback API when WebSerial api is not available. It uses bundled javascript WebUSB FTDI Driver v0.01a from: https://github.com/pmp-p/webusb-ftdi/blob/master/webusb-ftdi.js

Write support with WebUSB is now disabled - need test.
Write support with WebSerial is not tested. I need secondary usb-to-serial, to check if it works.

### CODE REVIEW

**Comment by slimonslimon** [comment]
FIY @kaklik

**Comment by Phoenix1747** [comment]
Please create the pull request for the `dev` branch not `main`. Thanks!

**Comment by Phoenix1747** [comment]
Did some changes to the code over at your fork: https://github.com/UniversalScientificTechnologies/Gamma-MCA/pull/2/

Please merge and I'll have a look at it once again. Then I'll probably merge as well. The Web Serial stuff all still works, so if the Web USB works on your end, great job!

**Review by Phoenix1747 — CHANGES_REQUESTED** [review]
Waiting for https://github.com/UniversalScientificTechnologies/Gamma-MCA/pull/2/ merge.

File: source/serial.ts [hunk 1]
```diff
@@ -37,49 +194,41 @@ export class SerialManager {
...
   async hideConsole(): Promise<void> {
-    if (!this.recording || !this.onlyConsole) return // Not recording or currently in a measurement so don't do anything...
+    if (!this.recording || !this.onlyConsole) return; // Not recording or currently in a measurement so don't do anything...
 
     this.onlyConsole = false;
     this.recording = false;
 
     try {
-      this.reader?.cancel();
+      this.port.close();
```
Comment (Phoenix1747) [inline source/serial.ts:231]: Please add an await here, just to be sure. I forgot that.

File: source/serial.ts [hunk 2]
```diff
@@ -98,7 +247,7 @@ export class SerialManager {
     this.timeDone += performance.now() - this.startTime;
 
     try {
-      this.reader?.cancel();
+      this.port.close();
```
Comment (Phoenix1747) [inline source/serial.ts:250]: Please add an await here too, just to be sure.

**Comment by slimonslimon** [comment]
Fix WebSerial Write means Fix WebUSB Write... sorry

**Comment by slimonslimon** [comment]
I tested folowing setups using two USBToSerial devices:

PC terminal <-> PC OpenMCA (WebSerial)
PC terminal <-> PC OpenMCA (WebUSB - using modified main.ts - disable WebSerial)
PC terminal <-> Android OpenMCA (WebUSB)
PC OpenMCA (WebSerial) <-> Android OpenMCA(WebUSB)

**Comment by Phoenix1747** [comment]
And did everything work?

**Comment by slimonslimon** [comment]
In my opinion yes.

But there are two things to discuse:
  - Send function consumes string, read functiuon produce Uint8Array - it's okey? 
  - Send function add '\n'. as End of Line, but parser of incoming data expected '\r\n' as End of Line

**Comment by Phoenix1747** [comment]
> Send function consumes string, read functiuon produce Uint8Array - it's okey?

Think that's fine. The only thing that's important is that both WebSerial and WebUSBSerial return the same type. The remaining conversion to strings is the same for both, so the SerialManager can handle that.

> Send function add '\n'. as End of Line, but parser of incoming data expected '\r\n' as End of Line

I think it would actually be better to remove the `\r` from the EOL. It just needs to use `\n` actually. I'll need to have a look at the serial data reduction to catch cases where either `\r\n` and `\n` are received. I'll do that once this PR is finished. (Some whitespace trimming fun)

**Review by slimonslimon — COMMENTED** [review]
shut i press this? (PR was accepted before)

**Comment by slimonslimon** [comment]
I still see changes requesed, but i think that I accept all all changes by your PR...

**Review by Phoenix1747 — APPROVED** [review]

**Comment by Phoenix1747** [comment]
Yeah, don't worry. Nice job.

**Comment by Phoenix1747** [comment]
@slimonslimon Can you please give the current `dev` branch another try with some WebUSB clients? I actually don't have any FTDI serial devices here, so can't test it unfortunately.

Had to fix some problems when multiple devices were connected and/or disconnected at the same time while recording. I haven't touched the WebUSBSerial class much, but just to be sure that the checks in `main.ts` didn't break something.

When this is working I will push the current `dev` branch to main, create a version tag and the new feature will be public.

**Comment by slimonslimon** [comment]
Im sorry but, current dev doesnt work properly with WebUSBSerial. 

First think i found: comment this line: 
https://github.com/OpenGammaProject/Gamma-MCA/blob/dev/source/serial.ts#L34
cause that it makes infinite loop without "await", so javascript cannot do any other "task" than this. (As far as I understand javascript) (when device not sending data - our device sends data about 1time per 10 seconds)

But i stil doesnt work when i remove comment, and I dont see other problem directly.

**Comment by Phoenix1747** [comment]
> cause that it makes infinite loop without "await", so javascript cannot do any other "task" than this. (As far as I understand javascript) (when device not sending data - our device sends data about 1time per 10 seconds)

Okay I see. That's something different with WebSerial then. Because WebSerial works without any delays. I will uncomment your delay in WebUSBSerial.

Is it more of a problem with finding and selecting the right port, or is there an issue with the recording on your end?

**Comment by slimonslimon** [comment]
The issue was between keyboard and chair.... 

When I want use (for dev purpose) the WebUSB on PC (where chromium has debug menus), I should remove kernel module whose block the usb device... I forget this...

**Comment by Phoenix1747** [comment]
Ok cool, thanks for testing! Will bump it to `main` then (with the uncommented line from above).

**Comment by slimonslimon** [comment]
Well... I tested it with two devices and PC with blocked WEBSerial API (one is USB2Serial which dont have opposite connection now, and second is LabDOS which produces data). When I have open GammaMCA-console with LABDOS and remove USB2Serial from USB, console keep recieving data. 

GammaMCA.console works on Android. 

So I thing you can publis it, because Im not able to setup more complicated experiment now. I will test it deeper on next week, if it is needed.

**Comment by Phoenix1747** [comment]
Yeah, all the WebSerial things still work just fine. So if at least _something_ works for now (even rudimentary) with WebUSB, then it's a clear win. We can check more complicated scenarios in the future anyways.

**Comment by Phoenix1747** [comment]
Just tested it with a random, new USB UART converter and made some minor adjustments to the filter so that it accepts _any_ FTDI chip, not only one productId. Works great now, tested with an Arduino Uno as a dummy, even the console shows fine.

**Pull request merged**